FragmentedSharedBuffer should always be immutable.

jyavenard opened the pull request:
#### 52383c657de22cc521391fd28d9a3fb97b283281
<pre>
FragmentedSharedBuffer should always be immutable.
<a href="https://bugs.webkit.org/show_bug.cgi?id=296798">https://bugs.webkit.org/show_bug.cgi?id=296798</a>
<a href="https://rdar.apple.com/157282731">rdar://157282731</a>

Reviewed by Gerald Squelart.

Following this change, a FragmentedSharedBuffer is now always immutable
and can never be modified once created.

The SharedBufferBuilder is the object where DataSegmentVector is now kept
and updated.

Upon call to SharedBufferBuilder::take or get, the SharedBuffer or FragmentedSharedBuffer
shared buffer will be lazily created.
Due to how ScriptBuffer uses a SharedBufferBuilder::get() we have to keep this
SharedBuffer alive until the next time a segment is added (or cleared).
This dangerous will be rectified in a follow-up change.

Covered by existing API tests and more added to broaden coverage.

Canonical link: <a href="https://commits.webkit.org/298282@main">https://commits.webkit.org/298282@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/10a46ee0cdcb986aa71391c61b16b83da1f9d4be

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/114994 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/34710 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/25191 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/121114 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/65644 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/374f8704-8546-4afc-b99e-ba59194eb0ab) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/116883 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/35355 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/43271 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/87380 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/42204 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/90d94970-43c4-4b5b-aed9-ac0fcc42b203) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/117942 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/28171 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/103240 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/67776 "Passed tests") | | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/27343 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/21359 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/64766 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/97552 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/21476 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/124304 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/41967 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/31370 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/96185 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/42339 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/99430 "Build is in progress. Recent messages:OS: Sonoma (14.7.3), Xcode: 15.4; Running apply-patch; Checked out pull request; Skipped layout-tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/95970 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24422 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/41162 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/19002 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/37986 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/41842 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/47377 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/41386 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/44703 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/43124 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/3cb7f90b5f49818becdbf00cc1ea44bdbaeec4a0. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/48941

squelart commented on Source/WebCore/platform/SharedBuffer.cpp:
```diff
@@ -652,15 +626,97 @@ Ref<SharedBuffer> SharedBufferBuilder::takeAsContiguous()
-void SharedBufferBuilder::ensureBuffer(size_t segments)
+void SharedBufferBuilder::ensureBuffer() const
 {
-    if (!m_buffer)
-        m_buffer = segments > 1 ? FragmentedSharedBuffer::create() : static_cast<Ref<FragmentedSharedBuffer>>(SharedBuffer::create());
+    if (!m_bufferStale)
+        return;
+    m_bufferStale = false;
+    if (m_isNull)
+        return;
+    m_buffer = createBuffer();
+}
```
I find it hard to reason! This seems to be the most important function.
Some thoughts:
- The name "ensureBuffer" typically means that we'll end up with a buffer, but there is at least one path (m_isNull) where this doesn't happen; also this function removes the staleness. Should the name be more like "ensureFreshBufferOrNull"? (If not, I may still be misunderstanding the logic!)
- There's a complex (to me) interaction between m_bufferStale, m_isNull, m_buffer, and m_segments. Would it be possible to use a state enum instead?

squelart reviewed (approved):
I think it should work as expected -> tentative r+, assuming all tests pass.
But I'm not 100% that I (or you) have thought of all possible states and transitions, so please consider my comment, see if things can be clarified a bit more, thanks.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/10a46ee0cdcb986aa71391c61b16b83da1f9d4be. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/48941

jyavenard committed:
FragmentedSharedBuffer should always be immutable.
https://bugs.webkit.org/show_bug.cgi?id=296798
rdar://157282731

Reviewed by Gerald Squelart.

Following this change, a FragmentedSharedBuffer is now always immutable
and can never be modified once created.

The SharedBufferBuilder is the object where DataSegmentVector is now kept
and updated.

Upon call to SharedBufferBuilder::take or get, the SharedBuffer or FragmentedSharedBuffer
shared buffer will be lazily created.
Due to how ScriptBuffer uses a SharedBufferBuilder::get() we have to keep this
SharedBuffer alive until the next time a segment is added (or cleared).
This dangerous will be rectified in a follow-up change.

Covered by existing API tests and more added to broaden coverage.

Canonical link: https://commits.webkit.org/298282@main
